src/utilities/extendSchema.js

```diff
@@ -275,30 +273,32 @@ export function extendSchema(
     const existingDirectives = schema.getDirectives();
     invariant(existingDirectives, 'schema must have default directives');
 
-    const newDirectives = directiveDefinitions.map(directiveNode =>
-      definitionBuilder.buildDirective(directiveNode),
+    return existingDirectives.concat(
+      directiveDefinitions.map(node => astBuilder.buildDirective(node)),
     );
-    return existingDirectives.concat(newDirectives);
   }
 
-  function getTypeFromDef<T: GraphQLNamedType>(typeDef: T): T {
-    const type = definitionBuilder.buildType(typeDef.name);
-    return (type: any);
+  function getExtendedType<T: GraphQLNamedType>(type: T): T {
+    if (!extendTypeCache[type.name]) {
+      extendTypeCache[type.name] = extendType(type);
+    }
+    return (extendTypeCache[type.name]: any);
   }
 
-  // Given a type's introspection result, construct the correct
-  // GraphQLType instance.
-  function extendType(type: GraphQLNamedType): GraphQLNamedType {
-    if (isObjectType(type)) {
-      return extendObjectType(type);
-    }
-    if (isInterfaceType(type)) {
-      return extendInterfaceType(type);
-    }
-    if (isUnionType(type)) {
-      return extendUnionType(type);
+  // Should be called only once per type so only getExtendedType should call it.
+  function extendType<T: GraphQLNamedType>(type: T): T {
+    let extendedType = type;
+    if (!isIntrospectionType(type)) {
+      if (isObjectType(type)) {
+        extendedType = extendObjectType(type);
+      } else if (isInterfaceType(type)) {
+        extendedType = extendInterfaceType(type);
+      } else if (isUnionType(type)) {
+        extendedType = extendUnionType(type);
+      }
     }
-    return type;
+    // Workaround: Flow should figure out correct type, but it doesn't.
+    return (extendedType: any);
```

Comment:
Rather than mutating a variable in each branch, just return the value in each branch. Flow can't figure out what the type of `extendedType` is, since there are so many different possible local mutations

Reply:
@leebyron I did that initially but it forces me to do typecast in every branch. Plus explaining workaround is more complicated.

> Flow can't figure out what the type of extendedType is, since there are so many different possible local mutations

Flow has exactly the same problem figuring types both with or without mutations. You can see it in CI logs for your commit and in Flow playground:
![image](https://user-images.githubusercontent.com/8336157/36050031-8cea7538-0ded-11e8-931d-f45cbb105ff6.png)
![image](https://user-images.githubusercontent.com/8336157/36049981-666b2c90-0ded-11e8-88ef-b2cc44921ed3.png)